Modules/Devices/src/Devices.Infrastructure/PushNotifications/DirectPush/ApplePushNotificationService/ApplePushNotificationServiceConnector.cs

```diff
@@ -60,20 +64,23 @@ public void ValidateRegistration(PnsRegistration registration)
             throw new InfrastructureException(InfrastructureErrors.InvalidPushNotificationConfiguration(_options.GetSupportedBundleIds()));
     }
 
-    private async Task HandleResponse(HttpResponseMessage response, string handle)
+    private async Task HandleResponse(HttpResponseMessage response, PnsRegistration registration)
     {
         if (response is { IsSuccessStatusCode: false })
         {
-            var responseContent = await response.Content.ReadAsStringAsync();
-            if (!responseContent.IsNullOrEmpty())
+            var responseContent = JsonConvert.DeserializeObject<ErrorResponse>(await response.Content.ReadAsStringAsync());
+            if (!responseContent.Reason.IsNullOrEmpty())
             {
                 _logger.LogError(
-                    "The following error occurred while trying to send the notification for handle '{handle}': '{responseContent}'",
-                    handle, responseContent);
+                    "The following error occurred while trying to send the notification for deviceId '{deviceId}': '{error}'",
+                    registration.DeviceId, responseContent.Reason);
+
+                if (responseContent.Reason == "Unregistered")
```

Comment:
This is the only possible reason we need to consider here?

Reply:
The documentation is not very specific on this topic (https://developer.apple.com/documentation/usernotifications/setting_up_a_remote_notification_server/handling_notification_responses_from_apns), but from my research this is the conclusion I arrived at.